client/gui/CCanvas.h

```diff
@@ -0,0 +1,53 @@
+/*
+ * CCanvas.h, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#pragma once
+
+#include "Geometries.h"
+
+struct SDL_Color;
+struct SDL_Surface;
+class IImage;
+enum EFonts : int;
+
+/// Class that represents surface for drawing on
+class CCanvas
+{
+	SDL_Surface * surface;
+public:
+
+	/// constructs canvas using existing surface. Caller maintains ownership on the surface
+	CCanvas(SDL_Surface * surface);
+
+	/// constructs canvas of specified size
+	CCanvas(const Point & size);
+
+	~CCanvas();
+
+	/// renders image onto this canvas at specified position
+	void draw(std::shared_ptr<IImage> image, const Point & pos);
+
+	/// renders section of image bounded by sourceRect at specified position
+	void draw(std::shared_ptr<IImage> image, const Point & pos, const Rect & sourceRect);
+
+	/// renders another canvas onto this canvas
+	void draw(std::shared_ptr<CCanvas> image, const Point & pos);

```

Comment:
IImage is always used while wrapped in shared_ptr. 
Using shared_ptr here makes more sense for consistency. 
Sure, I can switch it to reference but:
1) this means more boilerplate code in callers, since you'll need to dereference shared_ptr in every call
2) more error prone since it becomes caller responsibility to check for null = more points of failure.

I'd rather add nullptr checks inside CCanvas code, probably with asserts for debug code.